File: Week 03/id_738/LeetCode_200_738.py
class Solution(object):
    def numIslands(self, grid):
        """
        岛屿数量：https://leetcode-cn.com/problems/number-of-islands/
        
        :type grid: List[List[str]]
        :rtype: int
        """
        # 导弹轰炸法，每遇到一个岛屿，dfs法炸平整个岛
        # 到最后看看用了几个导弹就是有几个岛屿
        def dfs(x, y):
            if grid[x][y] == '0':
                return 0
            grid[x][y] = '0'
            for i in range(len(dx)):
                new_x, new_y = x + dx[i], y + dy[i]
                if 0 <= new_x < len(grid) and 0 <= new_y < len(grid[0]):
                    dfs(new_x, new_y)
            return 1
        dx = [-1, 1, 0, 0]
        dy = [0 ,0 ,-1, 1]
        islands = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == '1': 
                    islands += dfs(i, j)
        return islands

        

File: Week 06/id_738/pratice/LeetCode_200_738.java
public class Solution {
    // 岛屿数量：https://leetcode-cn.com/problems/number-of-islands/submissions/

    class UnionFind{
        private int[] parent;
        // 并查集有多少个组
        private int group = 0;
        UnionFind(int count) {
            //最后一个元素为特殊元素，所有水都是和他属于同个组
            parent = new int[count + 1];
            group = count + 1;
            for (int i = 0; i < parent.length; i++) {
                parent[i] = i;
            }
        }
        int find(int i) {
            int root = parent[i];
            while (root != parent[root]) {
                root = parent[root];
            }
            // 路径压缩
            while (i != parent[i]) {
                int tmp = parent[i];
                parent[i] = root;
                i = tmp;
            }
            return root;
        }
        void union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            if (rootX != rootY) {
                parent[rootY] = rootX;
                group --;
            }
        }
        int countGroup() {
            return group;
        }
    }

    int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) return 0;
        int m = grid.length;
        int n = grid[0].length;
        UnionFind uf = new UnionFind(m * n);
        int[] dx = new int[]{0, 0, 1, -1};
        int[] dy = new int[]{1, -1, 0, 0};
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    for (int di = 0; di < 4; di ++) {
                        int new_i = i + dy[di];
                        int new_j = j + dx[di];
                        if (0 <= new_i && new_i < m && 0 <= new_j && new_j < n && grid[new_i][new_j] == '1') {
                            uf.union(i * n + j, new_i * n + new_j);
                        }
                    }
                } else {
                    uf.union(i * n + j, m * n);
                }
            }
        }
        // 并查集有多少个组就是多少个岛屿
        return uf.countGroup() - 1;
    }
}